### PR TITLE
Examples for new feature webinar

### PR DESCRIPTION
**Pull request by sitvanit83**

### CODE REVIEW

File: CVLByExample/ERC20/contracts/broken/ERC20.sol [hunk 1]
```diff
@@ -407,4 +407,18 @@ contract ERC20 is IERC20, IERC20Metadata {
...
+        unchecked {
+            _balances[msg.sender] += amount;
+            _totalSupply += amount;
+        }
+    }
+
+    function add(uint256 x, uint256 y) external pure returns(uint256) {
+        unchecked {
+         return x + y;
+        }
```
Comment (nd-certora) [inline CVLByExample/ERC20/contracts/broken/ERC20.sol:422]: is this used? where is the unchecked?

**Review by nd-certora — CHANGES_REQUESTED** [review]

File: CVLByExample/ERC20/certora/specs/ERC20.spec [hunk 2]
```diff
@@ -132,3 +147,56 @@ rule sanity {
...
+    uint256 amount2;
+
+    storage initial = lastStorage;
+    depositAmount(e, amount1);
+    depositAmount(e, amount2);
+    storage afterTwoSteps = lastStorage;
+
+    depositAmount(e, assert_uint256(amount1 + amount2)) at initial;
+    storage afterOneStep = lastStorage;
+    assert afterOneStep == afterTwoSteps;
```
[the remark this answers is not in the document]
Comment (sitvanit83) [reply CVLByExample/ERC20/certora/specs/ERC20.spec:199]: moved

File: CVLByExample/ERC20/certora/specs/ERC20.spec [hunk 3]
```diff
@@ -93,6 +93,21 @@ rule onlyHolderCanChangeAllowance {
         "only approve and increaseAllowance can increase allowances";
 }
 
+rule onlyApproveIncreasesAllowance {
+    address holder; address spender;
+
+    mathint allowance_before = allowance(holder, spender);
+
+    method f; env e; calldataarg args; // was: env e; uint256 amount;
+    f(e, args);                        // was: approve(e, spender, amount);
```
Comment (nd-certora) [inline CVLByExample/ERC20/certora/specs/ERC20.spec:102]: remove the comments " was:... "

File: CVLByExample/ERC20/certora/specs/ERC20Fixed.spec [hunk 4]
```diff
@@ -117,3 +123,88 @@ rule sanity {
...
+    require balanceOf(e.msg.sender) > amount;
+    require e.msg.value == 0;
+    require balanceOf(recipient) + amount < max_uint;
+    require e.msg.sender != 0;
+    require recipient != 0;
+    deposit(e);
+    depositAmount(e, amount);
+    transfer@withrevert(e, recipient, amount);
+    satisfy totalSupply() > 0;  
+}
```
Comment (nd-certora) [inline CVLByExample/ERC20/certora/specs/ERC20Fixed.spec:160]: not clear what's the purpose of this rule, especially the withrevert

File: CVLByExample/ERC20/certora/specs/ERC20Fixed.spec [hunk 5]
```diff
@@ -117,3 +123,88 @@ rule sanity {
...
+
+// A non-vacuous example where transfer() does not revert.
+rule satisfyVacuityCorrection {
+    env e; address recip; uint amount;
+
+    require balanceOf(e.msg.sender) > 0;
+
+    transfer@withrevert(e, recip, amount);
+
+    satisfy balanceOf(e.msg.sender) == 0;
```
Comment (nd-certora) [inline CVLByExample/ERC20/certora/specs/ERC20Fixed.spec:172]: to show a case that transfer does not revert remove the withrevert

File: CVLByExample/ERC20/contracts/broken/ERC20.sol [hunk 6]
```diff
@@ -407,4 +407,13 @@ contract ERC20 is IERC20, IERC20Metadata {
         (bool success, ) = msg.sender.call{value: amount}("");
         require(success);
     }
+
+
+    function depositAmount(uint256 amount) external {
```
Comment (nd-certora) [inline CVLByExample/ERC20/contracts/broken/ERC20.sol:412]: this is not deposit, lets think of a different example that is more realistic, maybe 
depositTo(address to, uint256 amount)  payable and check the msg.value > amount and the rest goes to msg.sender

**Review by nd-certora — CHANGES_REQUESTED** [review]

File: CVLByExample/ERC20/certora/specs/ERC20.spec (hunk 3 above)
Comment (sitvanit83) [reply CVLByExample/ERC20/certora/specs/ERC20.spec:102]: removed

File: CVLByExample/ERC20/certora/specs/ERC20Fixed.spec (hunk 5 above)
Comment (sitvanit83) [reply CVLByExample/ERC20/certora/specs/ERC20Fixed.spec:172]: removed

File: CVLByExample/ERC20/contracts/broken/ERC20.sol (hunk 6 above)
Comment (sitvanit83) [reply CVLByExample/ERC20/contracts/broken/ERC20.sol:412]: The "overflow" rules want to demonstrate that using require_uint hides the overflow and it doesn't happen with the changes you suggest. I changed the name for these rules and added the rule depositTo  for other places it is used.

**Review by nd-certora — APPROVED** [review]

**Pull request merged**